Cleanup logs after successful test runs

dougbu opened the pull request:
- #39038
- add per-test `LoggedTestCleanerFixtureAttribute` to help track test failures
- add per-assembly `LoggedTestCleaner` to perform actual log directory Cleanup
- extend `AssemblyFixtureAttribute` and `AspNetTestAssemblyRunner` to support communication between the above
- use `LoggedTestCleanerFixtureAttribute` in test projects that produce large numbers of usually-useless log files

dougbu commented:
Working to address a couple of concerns:
- There are typos in two of the projects I changed.
- Cleanup doesn't happen if _any_ test in an assembly fails. Better if `LoggedTestBase` and any similar test or test base classes notified the cleaner when their specific tests fail. Will probably make `AssemblyTestLog` the cleaner (that'll also get its `Dispose()` method working).
- On the other hand, this cleanup doesn't happen for many of test classes that rely on `LoggedTestBase` let alone `AssemblyTestLog` or (even more broadly) otherwise write to `TestOutputDirectoryAttribute`-specified folders. Better to change Microsoft.AspNetCore.Testing.props than individual test projects although I may ignore the "otherwise write to `TestOutputDirectoryAttribute`-specified folders" part (by having tests indicate their interest in cleanup and assuming the two purposes don't overlap in a single test project).
- The `Instance` assumption is too fragile.

Will comment when I mark this PR ready for review again.

dougbu committed:
Cleanup logs after successful test runs
- #39038
- update `AssemblyTestLog` to perform actual log directory cleanup
  - add `ReportTestFailure()` method for tests to report failures, disabling cleanup
  - add `IAcceptFailureReports` for `AspNetTestAssemblyRunner` to report failures to `AssemblyTestLog`
- extend `AspNetTestAssemblyRunner` to optionally create fixture instances using `static ForAssembly(Assembly)`
- add `AssemblyTestLogFixtureAttribute` to register `AssemblyTestLog` as an assembly fixture
  - use `AssemblyTestLogFixtureAttribute` in all test projects
- disable log cleanup in three existing tests

nits:
- use `is [not] null` and `new()` more

halter73 commented:
> Cleanup doesn't happen if *any* test in an assembly fails.

I prefer this. Sometimes it's nice to look at similar passing tests to see if there are patterns in the logs. I hope most test assemblies pass at a high enough rate that we shouldn't bother optimizing the artifact size when they fail.

It'd really be nice if I could also look at logs from passing tests in other assemblies when there's a failure. Can we only clean logs if the entire test matrix succeeds? When I'm debugging rare ci-only test failures, I want to keep all the clues available to me.

dougbu commented on src/Testing/src/xunit/AspNetTestAssemblyRunner.cs:
```diff
@@ -33,24 +34,39 @@ protected override async Task AfterTestAssemblyStartingAsync()
-                var ctorWithDiagnostics = fixtureAttribute.FixtureType.GetConstructor(new[] { typeof(IMessageSink) });
                 object instance = null;
-                if (ctorWithDiagnostics != null)
+                var staticCreator = fixtureAttribute.FixtureType.GetMethod(
+                    name: "ForAssembly",
+                    bindingAttr: BindingFlags.Public | BindingFlags.Static,
+                    binder: null,
+                    types: new[] { typeof(Assembly) },
+                    modifiers: null);
```
This is specific to `AssemblyTestLog` currently but is a pattern other classes could follow. Adding a default constructor to `AssemblyTestLog` wouldn't be a good idea.

dougbu commented:
Ready for review

dougbu commented:
This does a great deal of cleanup compared to the before picture, especially when looking at non-Windows work items (none of which create more than 5 jobs). I may look into the IIS logging but that doesn't appear urgent.

### Before
[Build #20220228.5](https://dev.azure.com/dnceng/public/_build/results?buildId=1636034&view=results) of main.

WorkItemFriendlyName | FileCount
-- | --
Microsoft.AspNetCore.Server.Kestrel.Core.Tests--net7.0 | 3551
InMemory.FunctionalTests--net7.0 | 1289
Microsoft.AspNetCore.SignalR.Client.FunctionalTests--net7.0 | 810
Microsoft.AspNetCore.Http.Extensions.Tests--net7.0 | 395
Interop.FunctionalTests--net7.0 | 391
Microsoft.AspNetCore.SignalR.Tests--net7.0 | 261
IISExpress.FunctionalTests--net7.0 | 194
Microsoft.AspNetCore.SignalR.Client.Tests--net7.0 | 167
Microsoft.AspNetCore.Http.Connections.Tests--net7.0 | 150
IIS.LongTests--net7.0 | 114

### After

[Build #20220227.10](https://dev.azure.com/dnceng/public/_build/results?buildId=1635326&view=results) of this PR.

WorkItemFriendlyName | FileCount
-- | --
IISExpress.FunctionalTests--net7.0 | 190
IIS.LongTests--net7.0 | 113
IIS.FunctionalTests--net7.0 | 104
IIS.NewHandler.FunctionalTests--net7.0 | 84
IIS.NewShim.FunctionalTests--net7.0 | 82
IIS.ShadowCopy.Tests--net7.0 | 13
Microsoft.AspNetCore.Mvc.FunctionalTests--net7.0 | 10
Microsoft.Extensions.Features.Tests--net7.0 | 5
InteropTests--net7.0 | 5
Microsoft.Extensions.Diagnostics.HealthChecks.Tests--net462 | 5

### For posterity

<details>
<summary>Sample query</summary>

```
Files
| where JobName == "6d309ca5-777b-4aeb-9861-5cc4e4ba1213"
| summarize FileCount = count() by WorkItemFriendlyName
| top 10 by FileCount desc
```
</details>

dougbu commented:
/fyi @MattGal since you reported #39038

dougbu commented:
Hmm, have some room to improvement on the space our files consume. `FileSize` below adds all file sizes together for each work item.

WorkItemFriendlyName | TotalSize
-- | --
IIS.ShadowCopy.Tests--net7.0 | 440 MB
Microsoft.AspNetCore.SignalR.Client.FunctionalTests--net7.0 | 74 MB
IISExpress.FunctionalTests--net7.0 | 73 MB
Microsoft.AspNetCore.Server.Kestrel.Core.Tests--net7.0 | 51 MB
Sockets.FunctionalTests--net7.0 | 40 MB
InMemory.FunctionalTests--net7.0 | 30 MB
Microsoft.AspNetCore.Mvc.Core.Test--net7.0 | 25 MB
ProjectTemplates.Tests--net7.0 | 20 MB
Microsoft.AspNetCore.Routing.Tests--net7.0 | 19 MB
Microsoft.AspNetCore.Mvc.ViewFeatures.Test--net7.0 | 15 MB

dougbu commented:
Oops, that was for the build of 'main'. The after story (again, on Windows) is

WorkItemFriendlyName | TotalSize
-- | --
IIS.ShadowCopy.Tests--net7.0 | 147 MB
IISExpress.FunctionalTests--net7.0 | 69 MB
Microsoft.AspNetCore.SignalR.Client.FunctionalTests--net7.0 | 51 MB
Microsoft.AspNetCore.Server.Kestrel.Core.Tests--net7.0 | 47 MB
Sockets.FunctionalTests--net7.0 | 28 MB
Microsoft.AspNetCore.Mvc.Core.Test--net7.0 | 25 MB
InMemory.FunctionalTests--net7.0 | 23 MB
Microsoft.AspNetCore.Routing.Tests--net7.0 | 19 MB
Microsoft.AspNetCore.Mvc.ViewFeatures.Test--net7.0 | 15 MB
IIS.LongTests--net7.0 | 13 MB

BrennanConroy commented on src/Testing/src/AssemblyTestLog.cs:
```diff
@@ -238,10 +256,19 @@ public static AssemblyTestLog ForAssembly(Assembly assembly)
 
+    public void ReportTestFailure()
+    {
+        if (_cleanupState == CleanupState.CleanupRequested)
```
While I get what this is doing, the check isn't really needed and sort of makes it confusing when reading.
I would just always set the state, change the function name to something like "MarkHasTestFailure" or "RecordHasTestFailure", and if you really like the check maybe throw an exception.

dougbu replied:
Glad you pointed me here @BrennanConroy. My intention was to ensure we didn't lose track of reported test failures because another class requested cleanup _later_. This line means problems reported before the first cleanup request are ignored ☹️

Just updating `_cleanupState` is more correct but insufficient because other `AssemblyTestLog` users in the same assembly might not inherit from `LoggedTestBase`. I need to ensure all users report their failures if any users exist in the assembly because the cleanup has to be all or nothing.

dougbu replied:
I am working on two options here:

1. Revert changes to `LoggedTestBase` and have `AspNetTestAssemblyRunner` somehow notify `AssemblyTestLog` of test failures. This would entirely eliminate the concept of registering for cleanup and would mean _any_ failure in a test assembly prevents cleanup.
2. Update all `AssemblyTestLog` users to make sure they notify that class of test failures. Basically this option would make _not_ registering fail loudly and longly. If I go this route, I'll use @halter73's two-`bool` suggestion.

BrennanConroy reviewed (approved):
Overall looks good, no changes needed to existing tests, and looks like it works as expected from my quick glance at the quarantine and non-quarantine runs.

halter73 commented:
> I could leave things as-is, meaning (switching from InMemory.FunctionalTests to Kestrel.Core.Tests) a failure in `AddressBinderTests` or `AsciiDecodingTests` would cause logs to be saved for `ConnectionDispatcherTests` and other logged tests in that assembly. That seems a bit silly to me but can be persuaded if you prefer it.

Is this where we landed? As long as all the logs from the same helix job as the failing test are kept it's fine. I mostly like keeping the logs from other tests to find evidence of thread pool starvation, ephemeral port exhaustion or other similar environmental issues that can cause a bunch of tests in the same job to fail.

dougbu commented:
> Is this where we landed? As long as all the logs from the same helix job as the failing test are kept it's fine.

That was my proposal. I'm asking if it's enough for you.

I should have mentioned: Because Helix work items are per-assembly i.e. test a single TFM from a single test project, you'll get _all_ of the logs for the work item. For something like ServerComparison.FunctionalTests, a single test failure won't impact logging for other tests sharing the same build agent but you'll get all of the logs for that assembly.

dougbu commented on src/Testing/src/AssemblyTestLog.cs:
```diff
@@ -216,20 +217,37 @@ public static AssemblyTestLog Create(Assembly assembly, string baseDirectory)
-            if (!_logs.TryGetValue(assembly, out var log))
+            if (_logs.TryGetValue(assembly, out var log))
+            {
+                if (cleanupBaseDirectory && log._cleanupState == CleanupState.None)
+                {
+                    log._cleanupState = CleanupState.CleanupRequested;
```
Thinking about multiple test classes executing their tests in parallel, should this be bullet-proof e.g. using `Interlocked.CompareExchange(...)` (and a full-sized `int` `enum`)❔

halter73 replied:
You could always lock in `ReportTestFailure`. Although per-instance lock might be better than the static `_lock`. I think the best options is to avoid locking or interlocking entirely by turning `_cleanupState` into two bools: `bool _wasTestFailure` and `bool _cleanupRequested`.

Based on my reading, neither of these bools could ever transition from `true` to `false`, so there's no reason for any comparisons before setting either of these to `true`. `Dispose` can then delete the directory only if `_cleanupRequested && !_wasTestFailure`.

dougbu replied:
That'll work @halter73. I merged them into an `enum` primarily to make updates atomic but that didn't work out very well 😉

dougbu committed:
Add tests of new cleanup features
- also cover a few existing methods

nit: move `AssemblyTestLogTests` to same namespace as `AssemblyTestLog`

dougbu commented on src/Mvc/test/Mvc.FunctionalTests/ErrorPageTests.cs:
```diff
@@ -183,9 +183,4 @@ public async Task AggregateException_FlattensInnerExceptions()
-
-    public void Dispose()
-    {
-        _assemblyTestLog.Dispose();
```
This is important because individual tests don't own `AssemblyTestLog` instances, `AspNetTestAssemblyRunner` does

dougbu reviewed (commented):
I'm requesting re-reviews because this iteration is fairly different from what was in this PR before. A lot of that's in testing of Microsoft.AspNetCore.Testing but the src/ change to have `AspNetTestAssemblyRunner` handle failure reporting to `AssemblyTestLog` is significant too.

wtgodbe commented on src/Testing/test/TestableAssembly.cs:
```diff
@@ -0,0 +1,95 @@
+        var generator = method.GetILGenerator();
+        if (failTestCase)
+        {
+            // Assert.True(condition: false);
```
Remove?

wtgodbe replied:
Oh, never mind

dougbu committed:
!fixup! Make a new test more consistent

dougbu commented:
Below is from the console log of recent failure on Helix OSX agent, showing both the settings for Microsoft.AspNetCore.Testing, the `ForAssembly_ReturnsSameInstanceForSameAssembly()`'s `Disposable` instance, then the same for the inner dynamic assembly and fake test. Everything lines up but still the /var/folders/hv/54dwf7d169j4qnkvdmlkg0lw0000gq/T/TestLogging_8dcd6868d67c47558b55586a88470821/Testf6be736258cf46b2b2902121c279f9c8/net7.0 directory contains only the MyTestClass/MyTestClass.MyTestMethod.log file and no global.log. global.log files are created as expected in the other tests on both Linux and OSX.

What's worth checking next❔

``` text
 _assembly: 'Microsoft.AspNetCore.Testing.Tests, Version=42.42.42.42, Culture=neutral, PublicKeyToken=adb9793829ddae60'
 _baseDirectory: 'artifacts/log/Microsoft.AspNetCore.Testing.Tests/net7.0'
 _globalLogger: 'Microsoft.Extensions.Logging.Logger'
 _globalLogger.IsEnabled: 'True'
 _globalLoggerFactory: 'Microsoft.Extensions.Logging.LoggerFactory'
 _serviceProvider: 'Microsoft.Extensions.DependencyInjection.ServiceProvider'
 loggerProvider: Serilog.Extensions.Logging.SerilogLoggerProvider
 inner logger: 'Microsoft.Extensions.Logging.Logger'
 inner logger.IsEnabled: 'True'
 inner factory: 'Microsoft.Extensions.Logging.LoggerFactory'
 inner serviceProvider: 'Microsoft.Extensions.DependencyInjection.ServiceProvider'
 inner loggerProvider: Microsoft.Extensions.Logging.Testing.XunitLoggerProvider
 inner loggerProvider: Serilog.Extensions.Logging.SerilogLoggerProvider
 | [0.010s] Test Information: Created test log in /var/folders/hv/54dwf7d169j4qnkvdmlkg0lw0000gq/T/TestLogging_8dcd6868d67c47558b55586a88470821
 | [0.002s] TestLifetime Information: Starting test MyTestClass.MyTestMethod at 2022-03-17T01:13:02
 Starting test MyTestClass.MyTestMethod at 2022-03-17T01:13:02
 _assembly: 'Testf6be736258cf46b2b2902121c279f9c8, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
 _baseDirectory: '/var/folders/hv/54dwf7d169j4qnkvdmlkg0lw0000gq/T/TestLogging_8dcd6868d67c47558b55586a88470821/Testf6be736258cf46b2b2902121c279f9c8/net7.0'
 _globalLogger: 'Microsoft.Extensions.Logging.Logger'
 _globalLogger.IsEnabled: 'True'
 _globalLoggerFactory: 'Microsoft.Extensions.Logging.LoggerFactory'
 _serviceProvider: 'Microsoft.Extensions.DependencyInjection.ServiceProvider'
 loggerProvider: Serilog.Extensions.Logging.SerilogLoggerProvider
 inner logger: 'Microsoft.Extensions.Logging.Logger'
 inner logger.IsEnabled: 'True'
 inner factory: 'Microsoft.Extensions.Logging.LoggerFactory'
 inner serviceProvider: 'Microsoft.Extensions.DependencyInjection.ServiceProvider'
 inner loggerProvider: Microsoft.Extensions.Logging.Testing.XunitLoggerProvider
 inner loggerProvider: Serilog.Extensions.Logging.SerilogLoggerProvider
```

dougbu commented:
So, I finally checked if `AssemblyTestLog.ForAssembly(...)` produced global.log files and the answer is unfortunately **No**. That is, the `ForAssembly_ReturnsSameInstanceForSameAssembly()` failure in this PR is a demonstration of a real problem in Microsoft.AspNetCore.Testing on Linux and OSX. Raises the priority of figuring out what's going on. _Please help and / or provide suggestions❕_

dougbu committed:
Do cleanup before creating new logger
- was deleting the just-created global.log file

dougbu commented:
🎉 Found the problem 🎉

Please review the last commit if you're interested in the details. I'd also appreciate any comments before I squish 'n merge this sometime tomorrow&hellip;

BrennanConroy commented on src/Testing/src/AssemblyTestLog.cs:
```diff
@@ -218,12 +223,20 @@ public static AssemblyTestLog ForAssembly(Assembly assembly)
-                var baseDirectory = TestFileOutputContext.GetOutputDirectory(assembly);
+                var stackTrace = Environment.StackTrace;
+                if (!stackTrace.Contains(
+                    "Microsoft.AspNetCore.Testing"
```
Is this ever false? Wouldn't it include the current method which is in Microsoft.AspNetCore.Testing.

dougbu replied:
Thought it might be but you're probably right. I'll play with this in a follow-up when I get a bit of time.

dougbu commented:
/backport to release/6.0
